src/material/core/datetime/date-selection-model.ts

```diff
@@ -0,0 +1,207 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {FactoryProvider, Injectable, Optional, SkipSelf, OnDestroy} from '@angular/core';
+import {DateAdapter} from './date-adapter';
+import {Observable, Subject} from 'rxjs';
+
+/** A class representing a range of dates. */
+export class DateRange<D> {
+  constructor(
+    /** The start date of the range. */
+    readonly start: D | null,
+    /** The end date of the range. */
+    readonly end: D | null) {}
+}
+
+type ExtractDateTypeFromSelection<T> = T extends DateRange<infer D> ? D : NonNullable<T>;
+
+/** Event emitted by the date selection model when its selection changes. */
+export interface DateSelectionModelChange<S> {
+  /** New value for the selection. */
+  selection: S;
+
+  /** Object that triggered the change. */
+  source: any;
```

Comment:
Note that when we emit the change event we also attach which object triggered it. This isn't particularly clean, but we need to do it in order to avoid having things like the datepicker input reacting to changes that they themselves triggered.

Reply:
Yeah this makes sense, can we type it as `unknown` instead? I assume its just used for `===` comparisons? Do we also need some bit to tell us if the change is due to a user interaction or not? I was thinking that might be necessary to emit the right events to Angular Forms.